### PR TITLE
docker: set 5 seconds timeout on all queries

### PR DESCRIPTION
**Pull request by xvello**
If the docker daemon freezes, the agent currently can hang in the following components:
  - collector (docker corecheck frozen and blocking the check runner goroutine)
  - tagger
  - autodiscovery
  - live container collection

See https://github.com/DataDog/datadog-agent/issues/1487

**Users of the [docker-filter](https://github.com/DataDog/docker-filter) sidecar are not impacted, as idle connections are closed by haproxy after 5 seconds.**

### What does this PR do?

Add a timeout on every docker query. Default value is 5 seconds (same as agent5), configurable via the `docker_query_timeout` option.

- pass a `context.WithTimeout` to every call to the docker client lib
- move `pkg/util/ecs` to use `DockerUtil` instead of the bare client
- move `ConnectToDocker` that returns a bare docker client to private method. Only last use is [dead code in the process-agent](https://github.com/DataDog/datadog-process-agent/blob/7da4f5089728066e1928d34b1de713af1c6b9171/util/util.go#L88-L97)

We still have other naked `context.Background` uses we should address in followup-PRs:

- pkg/collector/providers/etcd.go
- pkg/logs/input/container/scanner.go & pkg/logs/input/container/docker.go (docker logs integration)

### Test images

- `datadog/agent-dev:xvello-docker-timeout`
- `datadog/agent-dev:xvello-docker-timeout-jmx`

### CODE REVIEW

**Comment by conorbranagan** [comment]
Timeouts look good. Feel free to kill `pkg/metadata/kubernetes/kubernetes.go` as we are not importing in the process-agent anymore after moving to the tagger.

File: pkg/util/docker/docker_util.go [hunk 1]
```diff
@@ -355,7 +366,9 @@ func (d *DockerUtil) GetHostname() (string, error) {
 // GetStorageStats returns the docker global storage stats if available
 // or ErrStorageStatsNotAvailable
 func (d *DockerUtil) GetStorageStats() ([]*StorageStats, error) {
-	info, err := d.cli.Info(context.Background())
+	ctx, cancel := context.WithTimeout(context.Background(), d.queryTimeout)
+	info, err := d.cli.Info(ctx)
+	cancel()
```
Comment (mfpierre) [inline pkg/util/docker/docker_util.go]: `defer cancel()` under context init

File: pkg/util/docker/docker_util.go [hunk 2]
```diff
@@ -412,7 +427,9 @@ func (d *DockerUtil) Inspect(id string, withSize bool) (types.ContainerJSON, err
 			return container, nil
 		}
 	}
-	container, _, err := d.cli.ContainerInspectWithRaw(context.Background(), id, withSize)
+	ctx, cancel := context.WithTimeout(context.Background(), d.queryTimeout)
+	container, _, err := d.cli.ContainerInspectWithRaw(ctx, id, withSize)
+	cancel()
```
Comment (mfpierre) [inline pkg/util/docker/docker_util.go]: `defer cancel()` under context init

File: pkg/util/docker/docker_util.go [hunk 3]
```diff
@@ -344,8 +345,18 @@ func (d *DockerUtil) Containers(cfg *ContainerListConfig) ([]*Container, error)
...
+	ctx, cancel := context.WithTimeout(context.Background(), d.queryTimeout)
+	defer cancel()
+	return d.cli.ContainerList(ctx, options)
+}
+
 func (d *DockerUtil) GetHostname() (string, error) {
-	info, err := d.cli.Info(context.Background())
+	ctx, cancel := context.WithTimeout(context.Background(), d.queryTimeout)
+	info, err := d.cli.Info(ctx)
+	cancel()
```
Comment (mfpierre) [inline pkg/util/docker/docker_util.go]: `defer cancel()` under context init

File: pkg/util/docker/docker_util.go [hunk 4]
```diff
@@ -372,7 +385,9 @@ func (d *DockerUtil) ResolveImageName(image string) (string, error) {
 	d.Lock()
 	defer d.Unlock()
 	if _, ok := d.imageNameBySha[image]; !ok {
-		r, _, err := d.cli.ImageInspectWithRaw(context.Background(), image)
+		ctx, cancel := context.WithTimeout(context.Background(), d.queryTimeout)
+		r, _, err := d.cli.ImageInspectWithRaw(ctx, image)
+		cancel()
```
Comment (mfpierre) [inline pkg/util/docker/docker_util.go]: `defer cancel()` under context init

File: pkg/util/docker/docker_util.go [hunk 5]
```diff
@@ -62,8 +58,12 @@ type DockerUtil struct {
 // init makes an empty DockerUtil bootstrap itself.
 // This is not exposed as public API but is called by the retrier embed.
 func (d *DockerUtil) init() error {
+	d.queryTimeout = config.Datadog.GetDuration("docker_query_timeout") * time.Second
+
 	// Major failure risk is here, do that first
-	cli, err := ConnectToDocker()
+	ctx, cancel := context.WithTimeout(context.Background(), d.queryTimeout)
+	cli, err := connectToDocker(ctx)
+	cancel()
```
Comment (mfpierre) [inline pkg/util/docker/docker_util.go]: `defer cancel()` under context init

File: pkg/util/docker/docker_util.go [hunk 6]
```diff
@@ -133,7 +128,9 @@ func ConnectToDocker() (*client.Client, error) {
 
 // Images returns a slice of all images.
 func (d *DockerUtil) Images(includeIntermediate bool) ([]types.ImageSummary, error) {
-	images, err := d.cli.ImageList(context.Background(), types.ImageListOptions{All: includeIntermediate})
+	ctx, cancel := context.WithTimeout(context.Background(), d.queryTimeout)
+	images, err := d.cli.ImageList(ctx, types.ImageListOptions{All: includeIntermediate})
+	cancel()
```
Comment (mfpierre) [inline pkg/util/docker/docker_util.go]: `defer cancel()` under context init

File: pkg/util/docker/docker_util.go [hunk 7]
```diff
@@ -478,7 +495,9 @@ func parseContainerHealth(status string) string {
 // AllContainerLabels retrieves all running containers (`docker ps`) and returns
 // a map mapping containerID to container labels as a map[string]string
 func (d *DockerUtil) AllContainerLabels() (map[string]map[string]string, error) {
-	containers, err := d.cli.ContainerList(context.Background(), types.ContainerListOptions{})
+	ctx, cancel := context.WithTimeout(context.Background(), d.queryTimeout)
+	containers, err := d.cli.ContainerList(ctx, types.ContainerListOptions{})
+	cancel()
```
Comment (mfpierre) [inline pkg/util/docker/docker_util.go]: `defer cancel()` under context init

File: pkg/util/docker/event_pull.go [hunk 8]
```diff
@@ -37,7 +36,9 @@ func (d *DockerUtil) openEventChannel(since, until time.Time, filter map[string]
 		Filters: queryFilter,
 	}
 
-	return d.cli.Events(context.Background(), options)
+	ctx, cancel := context.WithTimeout(context.Background(), d.queryTimeout)
```
Comment (mfpierre) [inline pkg/util/docker/event_pull.go]: no cancel?

**Review by mfpierre — CHANGES_REQUESTED** [review]
Looking good, a few things for consistency

File: pkg/util/docker/event_pull.go (hunk 8 above)
Comment (xvello) [reply pkg/util/docker/event_pull.go]: Canceling the requrest context is the user's (`LatestContainerEvents`) responsibility. Changing the method to ask for a `ctx` argument instead, for clearer ownership

**Review by mfpierre — DISMISSED** [review]
🐳 ⏲ ✅

**Review by mfpierre — APPROVED** [review]

**Pull request merged**